### PR TITLE
Desktop: Fixes #4332: Add Preview when using file:// protocol

### LINKED ISSUE #4332: No Preview, when using file:// (Only tested PDF)
## Environment

Joplin version: 1.5.14 (from AUR) and Pre-Release 1.6.6 (AppImage)
Platform: GNU/Linux (Desktop)
OS specifics: Archlinux

## Steps to reproduce

1. Attach a PDF-File to a note
2. Change the path from e.g. ``[test.pdf](:/631eaa8ceaf8449185096acf93ba9efd)`` to ``[test.pdf](file:///home/mika/test.pdf)``
3. See that the Link (opening it) is still working, but the preview disappeared

## Describe what you expected to happen
I expected to also see the preview, when I provide a full absolute path.

## Logfile
I found nothing interresting there (Only default output)
```
2021-01-10 19:17:32: "Loaded note:", "{"id":"28b7b9854548483397faa9fa6c0c5731","parent_id":"86d4a9426ed040f9acaced269c7aa868","title":"","body":"","created_time":1610302652342,"updated_time":1610302652342,"is_conflict":0,"latitude":"0.00000000","longitude":"0.00000000","altitude":"0.0000","author":"","source_url":"","is_todo":0,"todo_due":0,"todo_completed":0,"source":"joplin-desktop","source_application":"net.cozic.joplin-desktop","application_data":"","order":0,"user_created_time":1610302652342,"user_updated_time":1610302652342,"encryption_cipher_text":"","encryption_applied":0,"markup_language":1,"is_shared":0,"type_":1}"
2021-01-10 19:17:32: "CodeMirror: execCommand", "{"name":"editor.focus"}"
2021-01-10 19:17:34: "SearchEngine: Updating FTS table..."
2021-01-10 19:17:34: "Saving settings..."
2021-01-10 19:17:34: "Settings have been saved."
2021-01-10 19:17:34: "SearchEngine: Updated FTS table in 67ms. Inserted: 6. Deleted: 0"
2021-01-10 19:17:37: "CommandService::execute:", "attachFile", "[]"
2021-01-10 19:17:37: "CodeMirror: execCommand", "{"name":"attachFile"}"
2021-01-10 19:17:41: "Attaching /home/mika/test.pdf"
2021-01-10 19:17:41: "CommandService::execute:", "editor.focus", "[]"
2021-01-10 19:17:41: "CodeMirror: execCommand", "{"name":"editor.focus"}"
2021-01-10 19:17:41: "File was attached."
2021-01-10 19:17:42: "Saving note...", "{"id":"28b7b9854548483397faa9fa6c0c5731","parent_id":"86d4a9426ed040f9acaced269c7aa868","title":"test.pdf","body":"[test.pdf](:/631eaa8ceaf8449185096acf93ba9efd)"}"
2021-01-10 19:17:42: "Scheduling sync operation...", "30000"
2021-01-10 19:17:52: "SearchEngine: Updating FTS table..."
2021-01-10 19:17:52: "Saving settings..."
2021-01-10 19:17:52: "CommandService::execute:", "textCopy", "[]"
2021-01-10 19:17:52: "CodeMirror: execCommand", "{"name":"textCopy"}"
2021-01-10 19:17:52: "Settings have been saved."
2021-01-10 19:17:52: "SearchEngine: Updated FTS table in 71ms. Inserted: 1. Deleted: 0"
2021-01-10 19:17:54: "ResourceService::indexNoteResources: Start"
2021-01-10 19:17:54: "Saving settings..."
2021-01-10 19:17:54: "Settings have been saved."
2021-01-10 19:17:54: "ResourceService::indexNoteResources: Completed"
2021-01-10 19:17:54: "ResourceService::deleteOrphanResources:", "[]"
2021-01-10 19:17:56: "CommandService::execute:", "textCopy", "[]"
2021-01-10 19:17:56: "CodeMirror: execCommand", "{"name":"textCopy"}"
2021-01-10 19:18:04: "CommandService::execute:", "textCopy", "[]"
2021-01-10 19:18:04: "CodeMirror: execCommand", "{"name":"textCopy"}"
2021-01-10 19:18:11: "CommandService::execute:", "textCopy", "[]"
2021-01-10 19:18:11: "CodeMirror: execCommand", "{"name":"textCopy"}"
2021-01-10 19:18:12: "Preparing scheduled sync"
2021-01-10 19:18:12: "Synchroniser is missing credentials - manual sync required to authenticate."
2021-01-10 19:18:19: "CommandService::execute:", "textPaste", "[]"
2021-01-10 19:18:19: "CodeMirror: execCommand", "{"name":"textPaste"}"
2021-01-10 19:18:19: "Saving note...", "{"id":"28b7b9854548483397faa9fa6c0c5731","parent_id":"86d4a9426ed040f9acaced269c7aa868","title":"test.pdf","body":"[test.pdf](file:///home/mika/test.pdf)"}"
2021-01-10 19:18:19: "Scheduling sync operation...", "30000"
2021-01-10 19:18:21: "CommandService::execute:", "textCopy", "[]"
2021-01-10 19:18:21: "CodeMirror: execCommand", "{"name":"textCopy"}"
2021-01-10 19:18:29: "SearchEngine: Updating FTS table..."
2021-01-10 19:18:29: "Saving settings..."
2021-01-10 19:18:29: "Settings have been saved."
2021-01-10 19:18:29: "SearchEngine: Updated FTS table in 41ms. Inserted: 1. Deleted: 0"
2021-01-10 19:18:49: "Preparing scheduled sync"
2021-01-10 19:18:49: "Synchroniser is missing credentials - manual sync required to authenticate."
```

## Other
I also tried release 1.6.2, because it stated ``Fixed: file:// URLs would not be rendered correctly``, but it didn't work

### PR DESCRIPTION
**Pull request by Subhra264**
This pull request fixes #4332. The `linkReplacement.ts` inside `packages/renderer/MdToHtml` is updated to allow preview for files with `file://` protocol. The `@joplin/lib` package is added to the `renderer` package as a dependency to get the mime type of given file resource using the `mime-utils.js` module.

### CODE REVIEW

**Comment by tessus** [comment]
How to you turn this preview on/off? I do not want to see a preview of any local links.

**Comment by Subhra264** [comment]
Actually I didn't think about it yet, but the initial idea is that if some kind of check like `enablePreviewForLocalLinks` is passed through the options parameter to the `linkReplacement` function so that we can do :
```js
let resource = (enablePreviewForLocalLinks && isExternalFile && mime) ? { mime } : null;
// And similarly for the resourcePath
if (enablePreviewForLocalLinks && isExternalFile) {
	resourceFullPath = href;
}
```

As the preview totally depends on the values of `resource.mime` and the `resourcePath`. So, if the values of `resource.mime` and `resourcePath` can be set depending on the given `enablePreviewForLocalLinks` option, then I think it would work.

Not sure though, looking into it.

Thank you!

**Comment by tessus** [comment]
What I really don't want is that I can't turn this off. e.g. I don't use PDF nor video preview. And thus I certainly don't want to get a preview for any local file either.

File: packages/renderer/MdToHtml/linkReplacement.ts [hunk 1]
```diff
@@ -36,14 +37,15 @@ export default function(href: string, options: Options = null): LinkReplacementR
 
 	const resourceHrefInfo = urlUtils.parseResourceUrl(href);
 	const isResourceUrl = options.resources && !!resourceHrefInfo;
+	const isExternalFile = href.startsWith('file:///') || href.startsWith('file://');
```
Comment (roman-r-m) [inline packages/renderer/MdToHtml/linkReplacement.ts:40]: Doesn't the 2nd part in this condition includes the 1st?

**Comment by laurent22** [comment]
I don't really like this implementation as you're setting the resource variable even though it's not a resource, you have duplicate code to set the icon, etc. Ideally we work on refactoring and simplifying this function, not add hacks to cram whatever feature we feel like adding.

As I mentioned in the original issue, support for file:// is non-official, and if supporting PDF previewing means adding hacks, or require significant refactoring (to do it properly), then we just won't do it.

**Comment by Subhra264** [comment]
> As I mentioned in the original issue, support for file:// is non-official, and if supporting PDF previewing means adding hacks, or requires significant refactoring (to do it properly), then we just won't do it.

Sure, actually that's why I tried to keep the existing code as intact as possible.  But as you mentioned :

> You've also created a require cycle with lib => renderer => lib, which is likely to cause issues sooner or later.

lerna was also showing this kind of warning. I tried to avoid this, but couldn't find a way as I needed the mime-utils.js module.

> Ideally we should work on refactoring and simplifying this function, not add hacks to cram whatever feature we feel like adding.

Yes sir, got your point.

And thanks for the review...

File: packages/renderer/MdToHtml/linkReplacement.ts (hunk 1 above)
Comment (Subhra264) [reply packages/renderer/MdToHtml/linkReplacement.ts:40]: Sorry, my bad.

**Comment by laurent22** [comment]
Ok let's close it then, thanks for giving it a try @Subhra264 but in the end this feature is a bit out of scope.

**Pull request closed without merging**